File: ads_system/static/ads.js
/**
 * Sistema de Anúncios para Jogo WebGL Unity
 * 
 * Este script gerencia a exibição de anúncios no jogo:
 * - Banner rotativo no topo da página (1080×140px)
 * - Anúncios de tela cheia após game over (1080×1920px)
 * - Rastreamento de impressões e cliques
 */

// Configuração
const CONFIG = {
    // Intervalo de rotação do banner em milissegundos (7 segundos)
    BANNER_ROTATION_INTERVAL: 7000,
    
    // Número de game overs antes de mostrar anúncio de tela cheia
    GAMEOVER_COUNT_THRESHOLD: 5,
    
    // Duração do anúncio de tela cheia em milissegundos (5 segundos)
    FULLSCREEN_AD_DURATION: 5000,
    
    // URL da API para obter anúncios
    API_URL: window.location.hostname.includes('localhost') 
        ? 'http://localhost:5000/api'
        : 'https://seu-servidor-de-producao.com/api',
    
    // Seletor para o elemento do jogo Unity
    GAME_CONTAINER_SELECTOR: '#unity-container',
    
    // Seletor para o canvas do jogo Unity
    GAME_CANVAS_SELECTOR: '#unity-canvas'
};

// Estado do sistema de anúncios
const adSystem = {
    // Anúncios carregados
    bannerAds: [],
    fullscreenAds: [],
    
    // Índice do banner atual
    currentBannerIndex: 0,
    
    // Contador de game overs
    gameOverCount: 0,
    
    // Intervalo de rotação do banner
    bannerRotationInterval: null,
    
    // Elementos DOM
    bannerContainer: null,
    fullscreenContainer: null,
    
    // Instância do Unity
    unityInstance: null,
    
    // Flag para verificar se estamos na tela de jogo
    isInGameScreen: false,
    
    // Inicializa o sistema de anúncios
    init: function() {
        console.log('Inicializando sistema de anúncios...');
        
        // Criar elementos de contêiner
        this.createAdContainers();
        
        // Carregar anúncios do servidor
        this.loadAds();
        
        // Configurar comunicação com o Unity
        this.setupUnityConnection();
        
        // Verificar periodicamente se estamos na tela de jogo
        setInterval(() => this.checkGameScreen(), 1000);
    },
    
    // Cria os contêineres de anúncios
    createAdContainers: function() {
        // Criar contêiner de banner
        this.bannerContainer = document.createElement('div');
        this.bannerContainer.id = 'ad-banner-container';
        this.bannerContainer.style.cssText = `
            position: absolute;
            top: 0;
            left: 0;
            width: 100%;
            height: 140px;
            z-index: 1000;
            display: none;
            overflow: hidden;
        `;
        document.body.appendChild(this.bannerContainer);
        
        // Criar contêiner de anúncio de tela cheia
        this.fullscreenContainer = document.createElement('div');
        this.fullscreenContainer.id = 'ad-fullscreen-container';
        this.fullscreenContainer.style.cssText = `
            position: fixed;
            top: 0;
            left: 0;
            width: 100%;
            height: 100%;
            z-index: 2000;
            display: none;
            background-color: rgba(0, 0, 0, 0.8);
            justify-content: center;
            align-items: center;
        `;
        document.body.appendChild(this.fullscreenContainer);
    },
    
    // Carrega anúncios do servidor
    loadAds: async function() {
        try {
            // Carregar banners
            const bannerResponse = await fetch(`${CONFIG.API_URL}/ads/banner`);
            const bannerData = await bannerResponse.json();
            this.bannerAds = bannerData.ads || [];
            
            // Carregar anúncios de tela cheia
            const fullscreenResponse = await fetch(`${CONFIG.API_URL}/ads/fullscreen`);
            const fullscreenData = await fullscreenResponse.json();
            this.fullscreenAds = fullscreenData.ads || [];
            
            console.log(`Anúncios carregados: ${this.bannerAds.length} banners, ${this.fullscreenAds.length} tela cheia`);
            
            // Iniciar rotação de banners se houver anúncios
            if (this.bannerAds.length > 0) {
                this.startBannerRotation();
            }
        } catch (error) {
            console.error('Erro ao carregar anúncios:', error);
        }
    },
    
    // Configura comunicação com o Unity
    setupUnityConnection: function() {
        // Aguardar a instância do Unity ser criada
        window.addEventListener('unityInstance', (e) => {
            this.unityInstance = e.detail;
            console.log('Conexão com Unity estabelecida');
            
            // Adicionar função para o Unity chamar quando ocorrer game over
            window.gameOver = () => {
                this.handleGameOver();
            };
        });
    },
    
    // Verifica se estamos na tela de jogo
    checkGameScreen: function() {
        // Lógica para detectar se estamos na tela de jogo
        // Esta é uma implementação simplificada, você pode precisar ajustar
        // com base na estrutura específica do seu jogo
        
        // Exemplo: verificar se existe algum elemento específico da tela de jogo
        const isInGame = document.querySelector('.game-screen-indicator') !== null;
        
        // Alternativamente, o Unity pode enviar mensagens para indicar a tela atual
        
        // Se mudou de estado
        if (isInGame !== this.isInGameScreen) {
            this.isInGameScreen = isInGame;
            
            if (isInGame) {
                // Mostrar banner quando entrar na tela de jogo
                this.showBanner();
            } else {
                // Esconder banner quando sair da tela de jogo
                this.hideBanner();
            }
        }
    },
    
    // Inicia a rotação de banners
    startBannerRotation: function() {
        // Limpar intervalo existente
        if (this.bannerRotationInterval) {
            clearInterval(this.bannerRotationInterval);
        }
        
        // Mostrar o primeiro banner
        this.showCurrentBanner();
        
        // Configurar intervalo para rotação
        this.bannerRotationInterval = setInterval(() => {
            this.rotateToNextBanner();
        }, CONFIG.BANNER_ROTATION_INTERVAL);
    },
    
    // Mostra o banner atual
    showCurrentBanner: function() {
        if (this.bannerAds.length === 0) return;
        
        const ad = this.bannerAds[this.currentBannerIndex];
        
        // Limpar contêiner
        this.bannerContainer.innerHTML = '';
        
        // Criar elemento de banner
        const banner = document.createElement('div');
        banner.style.cssText = `
            width: 100%;
            height: 100%;
            background-image: url('${ad.imageUrl}');
            background-size: cover;
            background-position: center;
            cursor: pointer;
            position: relative;
        `;
        
        // Adicionar botão
        const button = document.createElement('a');
        button.href = ad.linkUrl;
        button.target = '_blank';
        button.style.cssText = `
            position: absolute;
            top: 50%;
            left: 50%;
            transform: translate(-50%, -50%);
            background-color: rgba(0, 123, 255, 0.7);
            color: white;
            border: none;
            padding: 8px 16px;
            border-radius: 4px;
            text-decoration: none;
            font-family: Arial, sans-serif;
        `;
        
        // Extrair domínio para o texto do botão
        try {
            const url = new URL(ad.linkUrl);
            const domain = url.hostname.replace('www.', '');
            button.textContent = `Visitar ${domain}`;
        } catch (e) {
            button.textContent = 'Visitar Site';
        }
        
        banner.appendChild(button);
        this.bannerContainer.appendChild(banner);
        
        // Registrar impressão
        this.recordImpression(ad.id, 'banner');
        
        // Adicionar evento de clique para rastreamento
        button.addEventListener('click', () => {
            this.recordClick(ad.id, 'banner');
        });
    },
    
    // Rotaciona para o próximo banner
    rotateToNextBanner: function() {
        if (this.bannerAds.length <= 1) return;
        
        this.currentBannerIndex = (this.currentBannerIndex + 1) % this.bannerAds.length;
        this.showCurrentBanner();
    },
    
    // Mostra o banner
    showBanner: function() {
        if (this.bannerAds.length === 0) return;
        this.bannerContainer.style.display = 'block';
    },
    
    // Esconde o banner
    hideBanner: function() {
        this.bannerContainer.style.display = 'none';
    },
    
    // Manipula evento de game over
    handleGameOver: function() {
        this.gameOverCount++;
        
        // Verificar se deve mostrar anúncio de tela cheia
        if (this.gameOverCount >= CONFIG.GAMEOVER_COUNT_THRESHOLD) {
            this.gameOverCount = 0; // Resetar contador
            this.showFullscreenAd();
        }
    },
    
    // Mostra anúncio de tela cheia
    showFullscreenAd: function() {
        if (this.fullscreenAds.length === 0) return;
        
        // Selecionar um anúncio aleatório
        const randomIndex = Math.floor(Math.random() * this.fullscreenAds.length);
        const ad = this.fullscreenAds[randomIndex];
        
        // Limpar contêiner
        this.fullscreenContainer.innerHTML = '';
        
        // Criar elemento de anúncio
        const adElement = document.createElement('div');
        adElement.style.cssText = `
            width: 100%;
            height: 100%;
            max-width: 1080px;
            max-height: 1920px;
            background-image: url('${ad.imageUrl}');
            background-size: contain;
            background-position: center;
            background-repeat: no-repeat;
            position: relative;
        `;
        
        // Adicionar botão
        const button = document.createElement('a');
        button.href = ad.linkUrl;
        button.target = '_blank';
        button.style.cssText = `
            position: absolute;
            bottom: 10%;
            left: 50%;
            transform: translateX(-50%);
            background-color: rgba(0, 123, 255, 0.7);
            color: white;
            border: none;
            padding: 12px 24px;
            border-radius: 4px;
            text-decoration: none;
            font-family: Arial, sans-serif;
            font-size: 18px;
            text-align: center;
            min-width: 200px;
        `;
        
        // Extrair domínio para o texto do botão
        try {
            const url = new URL(ad.linkUrl);
            const domain = url.hostname.replace('www.', '');
            button.textContent = `Visitar ${domain}`;
        } catch (e) {
            button.textContent = 'Visitar Site';
        }
        
        // Adicionar contador
        const counter = document.createElement('div');
        counter.style.cssText = `
            position: absolute;
            top: 20px;
            right: 20px;
            background-color: rgba(0, 0, 0, 0.5);
            color: white;
            padding: 5px 10px;
            border-radius: 20px;
            font-family: Arial, sans-serif;
        `;
        counter.textContent = '5';
        
        adElement.appendChild(button);
        adElement.appendChild(counter);
        this.fullscreenContainer.appendChild(adElement);
        
        // Mostrar anúncio
        this.fullscreenContainer.style.display = 'flex';
        
        // Pausar o jogo se possível
        if (this.unityInstance && typeof this.unityInstance.SendMessage === 'function') {
            this.unityInstance.SendMessage('GameManager', 'PauseGame');
        }
        
        // Registrar impressão
        this.recordImpression(ad.id, 'fullscreen');
        
        // Adicionar evento de clique para rastreamento
        button.addEventListener('click', () => {
            this.recordClick(ad.id, 'fullscreen');
        });
        
        // Iniciar contador regressivo
        let secondsLeft = 5;
        const countdownInterval = setInterval(() => {
            secondsLeft--;
            counter.textContent = secondsLeft;
            
            if (secondsLeft <= 0) {
                clearInterval(countdownInterval);
                this.hideFullscreenAd();
            }
        }, 1000);
        
        // Fechar automaticamente após o tempo definido
        setTimeout(() => {
            clearInterval(countdownInterval);
            this.hideFullscreenAd();
        }, CONFIG.FULLSCREEN_AD_DURATION);
    },
    
    // Esconde anúncio de tela cheia
    hideFullscreenAd: function() {
        this.fullscreenContainer.style.display = 'none';
        
        // Retomar o jogo se possível
        if (this.unityInstance && typeof this.unityInstance.SendMessage === 'function') {
            this.unityInstance.SendMessage('GameManager', 'ResumeGame');
        }
    },
    
    // Registra impressão de anúncio
    recordImpression: async function(adId, adType) {
        try {
            await fetch(`${CONFIG.API_URL}/ads/impression`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    adId: adId,
                    adType: adType
                })
            });
        } catch (error) {
            console.error('Erro ao registrar impressão:', error);
        }
    },
    
    // Registra clique em anúncio
    recordClick: async function(adId, adType) {
        try {
            await fetch(`${CONFIG.API_URL}/ads/click`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    adId: adId,
                    adType: adType
                })
            });
        } catch (error) {
            console.error('Erro ao registrar clique:', error);
        }
    }
};

// Inicializar sistema de anúncios quando a página carregar
document.addEventListener('DOMContentLoaded', () => {
    adSystem.init();
});

// Expor funções para o Unity chamar
window.showBanner = () => adSystem.showBanner();
window.hideBanner = () => adSystem.hideBanner();
window.gameOver = () => adSystem.handleGameOver();
window.showFullscreenAd = () => adSystem.showFullscreenAd();

File: models/ads.py
/**
 * Sistema de Anúncios para Jogos Unity WebGL
 * Versão: 2.0.0
 * 
 * Este script gerencia a exibição de banners (360x47px) e anúncios de tela cheia (360x640px)
 * em jogos Unity WebGL, com suporte a rastreamento de impressões e cliques.
 * 
 * Corrigido para garantir visibilidade dentro da área do jogo e compatibilidade com o backend.
 */

// Namespace isolado para evitar conflitos
const AdSystem = (function() {
    // Configurações
    const config = {
        apiBaseUrl: 'https://ads-system-backend.onrender.com',
        bannerEndpoint: '/api/banners',
        fullscreenEndpoint: '/api/fullscreen',
        impressionEndpoint: '/api/impression',
        clickEndpoint: '/api/click',
        bannerRotationInterval: 5000, // 5 segundos
        retryInterval: 10000, // 10 segundos
        debug: true
    };

    // Estado interno
    const state = {
        banners: [],
        fullscreenAds: [],
        currentBannerIndex: 0,
        bannerContainer: null,
        fullscreenContainer: null,
        bannerRotationTimer: null,
        retryTimer: null,
        unityInstance: null,
        unityCanvas: null,
        initialized: false,
        loading: false
    };

    // Sistema de logs
    const log = {
        info: function(message) {
            if (config.debug) {
                console.log(`[AdSystem ${new Date().toLocaleTimeString()}] ℹ️ ${message}`);
            }
        },
        warn: function(message) {
            if (config.debug) {
                console.log(`[AdSystem ${new Date().toLocaleTimeString()}] ⚠️ ${message}`);
            }
        },
        error: function(message) {
            if (config.debug) {
                console.log(`[AdSystem ${new Date().toLocaleTimeString()}] ❌ ${message}`);
            }
        }
    };

    // Utilitários
    const utils = {
        // Cria um elemento DOM com atributos
        createElement: function(tag, attributes = {}, styles = {}) {
            const element = document.createElement(tag);
            
            // Aplicar atributos
            Object.keys(attributes).forEach(key => {
                element.setAttribute(key, attributes[key]);
            });
            
            // Aplicar estilos
            Object.keys(styles).forEach(key => {
                element.style[key] = styles[key];
            });
            
            return element;
        },
        
        // Encontra o canvas do Unity
        findUnityCanvas: function() {
            // Tentar encontrar pelo ID padrão
            let canvas = document.getElementById('unity-canvas');
            
            // Se não encontrar pelo ID, procurar por qualquer canvas
            if (!canvas) {
                const canvases = document.getElementsByTagName('canvas');
                if (canvases.length > 0) {
                    canvas = canvases[0];
                }
            }
            
            // Se ainda não encontrou, procurar pelo container do Unity
            if (!canvas) {
                const unityContainer = document.getElementById('unity-container');
                if (unityContainer) {
                    const canvases = unityContainer.getElementsByTagName('canvas');
                    if (canvases.length > 0) {
                        canvas = canvases[0];
                    }
                }
            }
            
            return canvas;
        },
        
        // Faz uma requisição fetch com tratamento de erros
        fetchWithTimeout: async function(url, options = {}, timeout = 5000) {
            const controller = new AbortController();
            const timeoutId = setTimeout(() => controller.abort(), timeout);
            
            options.signal = controller.signal;
            
            try {
                const response = await fetch(url, options);
                clearTimeout(timeoutId);
                
                if (!response.ok) {
                    throw new Error(`HTTP error! Status: ${response.status}`);
                }
                
                return await response.json();
            } catch (error) {
                clearTimeout(timeoutId);
                throw error;
            }
        }
    };

    // Funções principais
    const api = {
        // Inicializa o sistema de anúncios
        init: function(unityInstance = null) {
            if (state.initialized) {
                log.warn('Sistema de anúncios já inicializado');
                return;
            }
            
            log.info('Inicializando sistema de anúncios...');
            
            // Armazenar referência ao Unity
            state.unityInstance = unityInstance;
            
            // Encontrar o canvas do Unity
            state.unityCanvas = utils.findUnityCanvas();
            if (!state.unityCanvas) {
                log.warn('Canvas do Unity não encontrado. Tentando novamente em 1 segundo...');
                setTimeout(() => api.init(unityInstance), 1000);
                return;
            }
            
            log.info('Canvas do Unity encontrado');
            
            // Criar containers para os anúncios
            api.createAdContainers();
            
            // Carregar anúncios
            api.loadBanners();
            api.loadFullscreenAds();
            
            state.initialized = true;
        },
        
        // Cria os containers para os anúncios
        createAdContainers: function() {
            // Criar container para banners
            if (!state.bannerContainer) {
                state.bannerContainer = utils.createElement('div', {
                    id: 'ad-banner-container'
                }, {
                    position: 'absolute',
                    top: '0',
                    left: '50%',
                    transform: 'translateX(-50%)',
                    width: '360px',
                    height: '47px',
                    zIndex: '9999',
                    display: 'none',
                    overflow: 'hidden',
                    backgroundColor: 'transparent'
                });
                
                // Anexar ao DOM
                if (state.unityCanvas && state.unityCanvas.parentElement) {
                    state.unityCanvas.parentElement.appendChild(state.bannerContainer);
                    log.info('Container de banner criado e anexado ao DOM');
                } else {
                    document.body.appendChild(state.bannerContainer);
                    log.warn('Container de banner anexado ao body (canvas parent não encontrado)');
                }
            }
            
            // Criar container para anúncios de tela cheia
            if (!state.fullscreenContainer) {
                state.fullscreenContainer = utils.createElement('div', {
                    id: 'ad-fullscreen-container'
                }, {
                    position: 'absolute',
                    top: '0',
                    left: '0',
                    width: '100%',
                    height: '100%',
                    zIndex: '10000',
                    display: 'none',
                    backgroundColor: 'rgba(0, 0, 0, 0.8)',
                    justifyContent: 'center',
                    alignItems: 'center'
                });
                
                // Anexar ao DOM
                if (state.unityCanvas && state.unityCanvas.parentElement) {
                    state.unityCanvas.parentElement.appendChild(state.fullscreenContainer);
                    log.info('Container de anúncio de tela cheia criado e anexado ao DOM');
                } else {
                    document.body.appendChild(state.fullscreenContainer);
                    log.warn('Container de anúncio de tela cheia anexado ao body (canvas parent não encontrado)');
                }
            }
        },
        
        // Carrega os banners do servidor
        loadBanners: function() {
            if (state.loading) return;
            state.loading = true;
            
            log.info('Carregando banners...');
            
            fetch(`${config.apiBaseUrl}${config.bannerEndpoint}`)
                .then(response => {
                    if (!response.ok) {
                        throw new Error('Resposta não ok');
                    }
                    return response.json();
                })
                .then(data => {
                    state.banners = data;
                    log.info(`${state.banners.length} banners carregados`);
                    
                    if (state.banners.length > 0) {
                        api.startBannerRotation();
                    }
                    
                    state.loading = false;
                })
                .catch(error => {
                    log.error(`Erro ao carregar banners: ${error.message}`);
                    state.loading = false;
                    
                    // Tentar novamente após um intervalo
                    if (!state.retryTimer) {
                        state.retryTimer = setTimeout(() => {
                            state.retryTimer = null;
                            api.loadBanners();
                        }, config.retryInterval);
                    }
                });
        },
        
        // Carrega os anúncios de tela cheia do servidor
        loadFullscreenAds: function() {
            log.info('Carregando anúncios de tela cheia...');
            
            fetch(`${config.apiBaseUrl}${config.fullscreenEndpoint}`)
                .then(response => {
                    if (!response.ok) {
                        throw new Error('Resposta não ok');
                    }
                    return response.json();
                })
                .then(data => {
                    state.fullscreenAds = data;
                    log.info(`${state.fullscreenAds.length} anúncios de tela cheia carregados`);
                })
                .catch(error => {
                    log.error(`Erro ao carregar anúncios de tela cheia: ${error.message}`);
                });
        },
        
        // Inicia a rotação de banners
        startBannerRotation: function() {
            if (state.bannerRotationTimer) {
                clearInterval(state.bannerRotationTimer);
            }
            
            // Mostrar o primeiro banner
            api.showNextBanner();
            
            // Iniciar rotação
            state.bannerRotationTimer = setInterval(() => {
                api.showNextBanner();
            }, config.bannerRotationInterval);
            
            log.info('Rotação de banners iniciada');
        },
        
        // Mostra o próximo banner na rotação
        showNextBanner: function() {
            if (state.banners.length === 0) {
                log.warn('Nenhum banner disponível para exibição');
                return;
            }
            
            // Avançar para o próximo banner
            state.currentBannerIndex = (state.currentBannerIndex + 1) % state.banners.length;
            const banner = state.banners[state.currentBannerIndex];
            
            // Mostrar o banner
            api.showBanner(banner);
        },
        
        // Mostra um banner específico
        showBanner: function(banner) {
            if (!state.bannerContainer) {
                log.error('Container de banner não inicializado');
                return;
            }
            
            // Limpar container
            state.bannerContainer.innerHTML = '';
            
            // Criar elemento de banner
            const bannerElement = utils.createElement('div', {
                class: 'ad-banner',
                'data-id': banner.id
            }, {
                width: '100%',
                height: '100%',
                cursor: 'pointer',
                position: 'relative'
            });
            
            // Criar imagem do banner
            const bannerImage = utils.createElement('img', {
                src: banner.imageUrl,
                alt: banner.title
            }, {
                width: '100%',
                height: '100%',
                objectFit: 'cover'
            });
            
            // Adicionar evento de clique
            bannerElement.addEventListener('click', () => {
                api.handleBannerClick(banner);
            });
            
            // Montar banner
            bannerElement.appendChild(bannerImage);
            state.bannerContainer.appendChild(bannerElement);
            
            // Exibir container
            state.bannerContainer.style.display = 'block';
            
            // Registrar impressão
            api.recordImpression(banner.id, 'banner');
            
            log.info(`Banner exibido: ${banner.title} (ID: ${banner.id})`);
        },
        
        // Mostra um anúncio de tela cheia
        showFullscreenAd: function() {
            if (state.fullscreenAds.length === 0) {
                log.warn('Nenhum anúncio de tela cheia disponível');
                return false;
            }
            
            // Selecionar um anúncio aleatório
            const randomIndex = Math.floor(Math.random() * state.fullscreenAds.length);
            const ad = state.fullscreenAds[randomIndex];
            
            if (!state.fullscreenContainer) {
                log.error('Container de anúncio de tela cheia não inicializado');
                return false;
            }
            
            // Limpar container
            state.fullscreenContainer.innerHTML = '';
            
            // Criar wrapper para centralizar o anúncio
            const adWrapper = utils.createElement('div', {
                class: 'ad-fullscreen-wrapper'
            }, {
                position: 'relative',
                width: '360px',
                height: '640px',
                margin: '0 auto',
                backgroundColor: '#fff',
                boxShadow: '0 0 20px rgba(0, 0, 0, 0.5)'
            });
            
            // Criar elemento de anúncio
            const adElement = utils.createElement('div', {
                class: 'ad-fullscreen',
                'data-id': ad.id
            }, {
                width: '100%',
                height: '100%',
                cursor: 'pointer',
                position: 'relative'
            });
            
            // Criar imagem do anúncio
            const adImage = utils.createElement('img', {
                src: ad.imageUrl,
                alt: ad.title
            }, {
                width: '100%',
                height: '100%',
                objectFit: 'cover'
            });
            
            // Criar botão de fechar
            const closeButton = utils.createElement('button', {
                class: 'ad-close-button',
                title: 'Fechar'
            }, {
                position: 'absolute',
                top: '10px',
                right: '10px',
                width: '30px',
                height: '30px',
                borderRadius: '50%',
                backgroundColor: 'rgba(0, 0, 0, 0.5)',
                color: '#fff',
                border: 'none',
                fontSize: '20px',
                cursor: 'pointer',
                display: 'flex',
                justifyContent: 'center',
                alignItems: 'center',
                zIndex: '1'
            });
            closeButton.innerHTML = '&times;';
            
            // Adicionar evento de clique no anúncio
            adElement.addEventListener('click', (e) => {
                // Ignorar clique se for no botão de fechar
                if (e.target !== closeButton && !closeButton.contains(e.target)) {
                    api.handleFullscreenAdClick(ad);
                }
            });
            
            // Adicionar evento de clique no botão de fechar
            closeButton.addEventListener('click', (e) => {
                e.stopPropagation();
                api.hideFullscreenAd();
            });
            
            // Montar anúncio
            adElement.appendChild(adImage);
            adElement.appendChild(closeButton);
            adWrapper.appendChild(adElement);
            state.fullscreenContainer.appendChild(adWrapper);
            
            // Exibir container
            state.fullscreenContainer.style.display = 'flex';
            
            // Registrar impressão
            api.recordImpression(ad.id, 'fullscreen');
            
            log.info(`Anúncio de tela cheia exibido: ${ad.title} (ID: ${ad.id})`);
            
            return true;
        },
        
        // Esconde o anúncio de tela cheia
        hideFullscreenAd: function() {
            if (state.fullscreenContainer) {
                state.fullscreenContainer.style.display = 'none';
                log.info('Anúncio de tela cheia fechado');
            }
        },
        
        // Trata o clique em um banner
        handleBannerClick: function(banner) {
            log.info(`Banner clicado: ${banner.title} (ID: ${banner.id})`);
            
            // Registrar clique
            api.recordClick(banner.id, 'banner');
            
            // Abrir URL de destino
            const targetUrl = banner.targetUrl || banner.linkUrl;
            if (targetUrl) {
                window.open(targetUrl, '_blank');
            }
        },
        
        // Trata o clique em um anúncio de tela cheia
        handleFullscreenAdClick: function(ad) {
            log.info(`Anúncio de tela cheia clicado: ${ad.title} (ID: ${ad.id})`);
            
            // Registrar clique
            api.recordClick(ad.id, 'fullscreen');
            
            // Abrir URL de destino
            const targetUrl = ad.targetUrl || ad.linkUrl;
            if (targetUrl) {
                window.open(targetUrl, '_blank');
            }
            
            // Fechar anúncio
            api.hideFullscreenAd();
        },
        
        // Registra uma impressão
        recordImpression: function(adId, adType) {
            fetch(`${config.apiBaseUrl}${config.impressionEndpoint}`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    adId: adId,
                    type: adType
                })
            })
            .then(response => {
                if (!response.ok) {
                    throw new Error('Resposta não ok');
                }
                return response.json();
            })
            .then(data => {
                log.info(`Impressão registrada: ${adType} ${adId}`);
            })
            .catch(error => {
                log.error(`Erro ao registrar impressão: ${error.message}`);
            });
        },
        
        // Registra um clique
        recordClick: function(adId, adType) {
            fetch(`${config.apiBaseUrl}${config.clickEndpoint}`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    adId: adId,
                    type: adType
                })
            })
            .then(response => {
                if (!response.ok) {
                    throw new Error('Resposta não ok');
                }
                return response.json();
            })
            .then(data => {
                log.info(`Clique registrado: ${adType} ${adId}`);
            })
            .catch(error => {
                log.error(`Erro ao registrar clique: ${error.message}`);
            });
        },
        
        // Mostra um anúncio de tela cheia após game over
        showGameOverAd: function() {
            return api.showFullscreenAd();
        },
        
        // Diagnóstico do sistema
        diagnose: function() {
            log.info('--- Diagnóstico do Sistema de Anúncios ---');
            log.info(`Inicializado: ${state.initialized}`);
            log.info(`Banners carregados: ${state.banners.length}`);
            log.info(`Anúncios de tela cheia carregados: ${state.fullscreenAds.length}`);
            log.info(`Container de banner existe: ${!!state.bannerContainer}`);
            log.info(`Container de anúncio de tela cheia existe: ${!!state.fullscreenContainer}`);
            log.info(`Canvas do Unity encontrado: ${!!state.unityCanvas}`);
            
            if (state.bannerContainer) {
                log.info(`Visibilidade do container de banner: ${state.bannerContainer.style.display}`);
                log.info(`Z-index do container de banner: ${state.bannerContainer.style.zIndex}`);
                log.info(`Posição do container de banner: top=${state.bannerContainer.style.top}, left=${state.bannerContainer.style.left}`);
            }
            
            return {
                initialized: state.initialized,
                bannersLoaded: state.banners.length,
                fullscreenAdsLoaded: state.fullscreenAds.length,
                bannerContainerExists: !!state.bannerContainer,
                fullscreenContainerExists: !!state.fullscreenContainer,
                unityCanvasFound: !!state.unityCanvas
            };
        },
        
        // Verifica e corrige a visibilidade dos containers
        checkVisibility: function() {
            if (!state.initialized) return;
            
            if (state.bannerContainer) {
                // Garantir que o container de banner esteja visível
                if (state.bannerContainer.style.display === 'none' && state.banners.length > 0) {
                    state.bannerContainer.style.display = 'block';
                    log.info('Visibilidade do container de banner corrigida');
                }
                
                // Garantir que o z-index seja alto o suficiente
                if (parseInt(state.bannerContainer.style.zIndex) < 9999) {
                    state.bannerContainer.style.zIndex = '9999';
                    log.info('Z-index do container de banner corrigido');
                }
            }
        }
    };

    // Iniciar verificação periódica de visibilidade
    setInterval(() => {
        api.checkVisibility();
    }, 5000);

    // Inicializar quando o documento estiver pronto
    if (document.readyState === 'complete' || document.readyState === 'interactive') {
        setTimeout(() => api.init(), 1000);
    } else {
        document.addEventListener('DOMContentLoaded', () => {
            setTimeout(() => api.init(), 1000);
        });
    }

    // Interface pública
    return {
        init: api.init,
        showBanner: api.showBanner,
        showFullscreenAd: api.showFullscreenAd,
        hideFullscreenAd: api.hideFullscreenAd,
        showGameOverAd: api.showGameOverAd,
        diagnose: api.diagnose
    };
})();

// Compatibilidade com Unity
function showBanner() {
    return AdSystem.showBanner();
}

function showFullscreenAd() {
    return AdSystem.showFullscreenAd();
}

function hideFullscreenAd() {
    return AdSystem.hideFullscreenAd();
}

function showGameOverAd() {
    return AdSystem.showGameOverAd();
}

// Inicializar quando o Unity estiver pronto
window.addEventListener('unityReady', function(e) {
    AdSystem.init(e.detail.unityInstance);
});

// Log de inicialização
console.log('[AdSystem] Sistema de anúncios carregado e pronto para inicialização');
